### PR TITLE
Support DTLS in TLS::Proxy.

### PR DESCRIPTION
**Pull request by fwh-dc**
- [x] tests are added or updated

This pull request adds DTLS support in the TLSProxy, and several of its dependencies to allow writing tests similar to sslrecords, sslvertol, and in the future test_tls13****.t.

The DTLS tests I have added so far is a limited set just to show how DTLS support in TLSProxy works. Of course there can bee added more tests in the future, but the main focus of this PR is to add the support to TLSProxy.

To avoid message fragmentation I have set the MTU size at 1500. I want the tests to be skipped if the test system does not support this, but I didn't know how to achieve that.

One thing to notice is that in order to test connection establishment failures we are currently waiting 10 secs for the test to time-out. I don't know if there's a more elegant way to do it.

### CODE REVIEW

**Comment by mattcaswell** [comment]
Oh....wow...this is fantastic!!

**Comment by mattcaswell** [comment]
CI failures look relevant BTW.

File: test/recipes/70-test_dtlsrecords.t [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+#Unrecognised record type tests
+
+#Test 3: Sending an unrecognised record type in DTLSv1.2 should fail
+$fatal_alert = 0;
+$proxy->clear();
+$proxy->serverflags("-min_protocol DTLSv1.2 -max_protocol DTLSv1.2");
+$proxy->clientflags("-max_protocol DTLSv1.2");
+$proxy->filter(\&add_unknown_record_type);
+ok($proxy->start() == 0, "Unrecognised record type in TLS1.2");
```
Comment (fwh-dc) [inline test/recipes/70-test_dtlsrecords.t:72]: TLS1.2 -> DTLS1.2

**Comment by fwh-dc** [comment]
> @fwh-dc even the default configurations seem to fail the tests, there should be nothing special needed to reproduce the failure: https://github.com/openssl/openssl/actions/runs/7552847775/job/20562487951?pr=23319
> 
> Hmm, maybe enable-fips is causing this?

I've done:
```
git clean -dffx && mkdir builddir && cd builddir && ../config && make -j test
```
and
```
git clean -dffx && mkdir builddir && cd builddir && ../config enable-fips && make -j test
```
Both are successfull and runs the test.

I am running on a mac, and I am wondering if it is something Linux specific that's causing the issue. Or a Perl version issue perhabs? I am using Perl 5.38.2

**Comment by mattcaswell** [comment]
I just tried it on my Linux machine and it worked just fine, so it doesn't seem to be something specific to Linux.

**Comment by mattcaswell** [comment]
Perhaps something timing related?

**Comment by fwh-dc** [comment]
Thanks for giving it a go. Could be timing. I'll dig in to it.

**Comment by t8m** [comment]
> @mattcaswell @t8m can I run a job on the CI without updating the PR?

IMO no.

**Comment by beldmit** [comment]
I think close/reopen PR will do the trick

**Comment by fwh-dc** [comment]
Here are some observations:

It seems like the connection from `s_client` to `TLSProxy` works.  

Next thing is that sending from `TLSProxy` to `s_server` fails with `ECONNREFUSED`. 

Most of the time (seems like there is some timing involved) `s_server `seems to stop abruptly before the `s_client` is even started. As seen by this trace:  

```
 Proxy started on port [::1]:51852
Server command: ../../util/wrap.pl ../../apps/openssl s_server -no_comp -engine ossltest -state -ext_cache -accept [::1]:52805 -cert ../../apps/server.pem -cert2 ../../apps/server.pem -naccept 1 -dtls -max_protocol DTLSv1.2 -mtu 1500 -cipher AES128-SHA -ciphersuites TLS_AES_128_GCM_SHA256 -min_protocol DTLSv1.2 -max_protocol DTLSv1.2
shutdown accept socket
shutting down SSL
CONNECTION CLOSED
   0 items in the session cache
   0 client connects (SSL_connect())
   0 client renegotiates (SSL_connect())
   0 client connects that finished
   0 server accepts (SSL_accept())
   0 server renegotiates (SSL_accept())
   0 server accepts that finished
   0 session cache hits
   0 session cache misses
   0 session cache timeouts
   0 callback cache hits
   0 cache full overflows (20480 allowed)
Engine "ossltest" set.
Using default temp DH parameters
ACCEPT
Server responds on [::1]:52805
```  

After tracing line numbers i found that it is this read that fails:
https://github.com/openssl/openssl/blob/5b2d8bc28a8ff59689da98f31459819db09a9099/apps/s_server.c#L2681-L2683

Another thing I noticed is that the log line 

```
 CONNECTED(00000003)
  ```
  
   from the `s_client` is always printed **after** the client is sending messages to `TLSProxy` on the CI. But on systems where the test passes the log line is printed **before** the client is sending messages to `TLSProxy`. Which suggests BIO_flush is called differently on the CI. 
   
   I don't know if those observations rings a bell to anyone?
   
   I've failed to find a way to reproduce this issue anywhere else than on the CI.
   
   Is there any setup on the CI that restricts UDP ports in some way?
Is there a test that show a simple s_server and s_client with dtls is able to communicate on the CI?
@mattcaswell @t8m

**Comment by mattcaswell** [comment]
>    I don't know if those observations rings a bell to anyone?

No bells ringing here.

>    Is there any setup on the CI that restricts UDP ports in some way?

Not that I am aware of.

> Is there a test that show a simple s_server and s_client with dtls is able to communicate on the CI?

There is a simple s_client/s_server test but it uses TLS not DTLS. Of course you might be able to adapt it as a test just to check it works.

https://github.com/openssl/openssl/blob/master/test/recipes/82-test_tfo_cli.t

**Comment by fwh-dc** [comment]
@mattcaswell 
That was a good advice. I did add a minimal test in https://github.com/openssl/openssl/pull/23357 and it passed on the CI. So I was looking at differences and one thing that catched me was the open() vs open2(). There was some (to me at least) hard to read handling of stdin/out that can be avoided with open2(). So I decided to refactor that part of the code. And it looks like that did the trick.

This PR should be ready for review.

**Comment by t8m** [comment]
It seems the tests get stuck on Windows.

**Comment by fwh-dc** [comment]
Agreed.. We are setting up a windows buildnode locally this week. I can test on that and find a fix. I'll ping you when I know more.

File: test/recipes/70-test_dtlsrecords.t [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    if disabled("dtls1_2");
+
+my $proxy = TLSProxy::Proxy->new_dtls(
+    undef,
+    cmdstr(app(["openssl"]), display => 1),
+    srctop_file("apps", "server.pem"),
+    (!$ENV{HARNESS_ACTIVE} || $ENV{HARNESS_VERBOSE})
+);
+
+plan tests => 4;
```
Comment (mattcaswell) [inline test/recipes/70-test_dtlsrecords.t:39]: For other TLSProxy tests we defer this until after the first `$proxy->start` call. The reason is that we had some problems on some odd platforms where the TLSProxy was not starting correctly. In the end we deferred the plan until after we knew the proxy was successfully started.

File: test/recipes/70-test_dtlsrecords.t [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    #Find ServerHello record and insert after that
+    my $i;
+    for ($i = 0; ${$proxy->record_list}[$i]->flight() < 1; $i++) {
+        next;
+    }
+    $i++;
+
+    splice @{$proxy->record_list}, $i, 0, $record;
+    $added_record = 1;
+}
```
Comment (mattcaswell) [inline test/recipes/70-test_dtlsrecords.t:153]: This is basically a reduced version of the sslrecords test - with essentially identical filters except for new->new_dtls. Can we not find a way to reuse the existing script for DTLS in order to avoid all the copy&paste?

File: util/perl/TLSProxy/Proxy.pm [hunk 4]
```diff
@@ -266,38 +309,6 @@ sub start
...
-    open(STDIN, "<&", $savedin);
-    close($savedin);
-
-    if (!defined($pid)) {
-        kill(3, $self->{real_serverpid});
-        die "Failed to capture s_server's output: $error\n";
-    }
-
-    $self->{serverpid} = $pid;
-
```
Comment (mattcaswell) [inline util/perl/TLSProxy/Proxy.pm:300]: Why is this no longer needed?

File: util/perl/TLSProxy/Proxy.pm [hunk 5]
```diff
@@ -436,15 +460,6 @@ sub clientstart
...
         $pid = $self->{serverpid};
-        print "Waiting for 'perl -ne print' process to close: $pid...\n";
-        $pid = waitpid($pid, 0);
-        if ($pid > 0) {
-            die "exit code $? from 'perl -ne print' process\n" if $? != 0;
-        } elsif ($pid == 0) {
-            kill(3, $self->{real_serverpid});
-            die "lost control over $self->{serverpid}?";
-        }
-        $pid = $self->{real_serverpid};
```
Comment (mattcaswell) [inline util/perl/TLSProxy/Proxy.pm:447]: Why is this deleted?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by fwh-dc** [comment]
@mattcaswell 

Just wanted to make sure everything passes here on the CI before replying to you.

Here is the current state: I’ve tested TLSProxy tests can run locally (mac, windows) and on the ci (linux, windows) I’ve tested DTLSProxy tests can run locally (mac) and on the ci (linux).

I spent last week debugging why it failed on the CI and later on windows. I am no socket/filehandle expert so these are just observations.

It turns out that each system seems to handle sockets and filehandles differently. I am suspecting that some systems (namely the github ci) close open udp sockets when the filehandle goes "out of scope" as is done here:
https://github.com/openssl/openssl/blob/0f644b96d209443b4566f7e86e3be2568292e75b/util/perl/TLSProxy/Proxy.pm#L290-L292

So to avoid that behaviour I refactored the existing code allowing the in/out filehandles for the s_server process to be kept open.  And then I figured, well I could also do it for the s_client process to be consistent.

Turns out that these changes makes the s_client/s_server hang on windows (but work on ci/macos) when terminating the test case (probably because something is keeping the socket open). Now if I remove the s_client process changes I can get the TLSProxy test cases to run on all platforms, but I haven’t found a way to run the DTLSProxy test cases on windows.   And I also found out that Windows don’t like chomp on the s_server output. So I have put that code back to its original state. 
I hope this state is reasonable to you. I will look at your suggestions now.  Please let me know your thoughts.

File: test/recipes/70-test_dtlsrecords.t (hunk 2 above)
Comment (fwh-dc) [reply test/recipes/70-test_dtlsrecords.t:39]: Cool, I'll update accordingly.

File: test/recipes/70-test_dtlsrecords.t (hunk 3 above)
Comment (fwh-dc) [reply test/recipes/70-test_dtlsrecords.t:153]: Possibly. I'm just not that familiar with the perl testing framework yet. You are thinking something in the terms of

```
my @run_with_dtls = (0, 1);

foreach(@run_with_dtls) {
    run_test($_)
}
```

The next question from me would be: What is a nice way to skip the DTLS tests on windows (Since I couldn't get it to work on windows)?

Comment (mattcaswell) [reply test/recipes/70-test_dtlsrecords.t:153]: Something along those lines.

> The next question from me would be: What is a nice way to skip the DTLS tests on windows (Since I couldn't get it to work on windows)?

You can test for windows via `if $^O eq 'MSWin32'`

You can skip individual tests like this:

https://github.com/openssl/openssl/blob/0f644b96d209443b4566f7e86e3be2568292e75b/test/recipes/70-test_key_share.t#L142-L151

Comment (fwh-dc) [reply test/recipes/70-test_dtlsrecords.t:153]: Cool, I'll give it a go

**Comment by mattcaswell** [comment]
> It turns out that each system seems to handle sockets and filehandles differently. 

This doesn't surprise me one bit. I recall in the past we have had similar difficulties getting things to work cross-platform.

> Turns out that these changes makes the s_client/s_server hang on windows (but work on ci/macos) when terminating the test case (probably because something is keeping the socket open). Now if I remove the s_client process changes I can get the TLSProxy test cases to run on all platforms, but I haven’t found a way to run the DTLSProxy test cases on windows. 

If we really can't figure out how to get it to work on Windows then just skipping DTLSProxy testing on Windows seems to be the way to go. Hopefully we can figure it out in time.

File: util/perl/TLSProxy/Proxy.pm (hunk 4 above)
Comment (fwh-dc) [reply util/perl/TLSProxy/Proxy.pm:300]: There's an explanation in the comments. I hope it answers this question.

**Comment by fwh-dc** [comment]
I've implemented the changes. Let me know if there's more.

**Comment by t8m** [comment]
The CI seems relevant. Also the Windows shared builds seem to be stuck? Not sure if this is caused by this PR or anything else?

**Comment by fwh-dc** [comment]
@mattcaswell Does not look good for windows when merging the tests together. Would it be acceptable to split the tests in tls and dtls specific test cases? ie. go back to [db80e6d](https://github.com/openssl/openssl/pull/23319/commits/db80e6d250d306fe54e8fffe6622f10ecd25daea).

**Comment by t8m** [comment]
Still some CI failures

**Comment by fwh-dc** [comment]
@mattcaswell @t8m I think I finally managed to get rid of the last few bugs. Let me know what you think.

**Comment by fwh-dc** [comment]
@mattcaswell @t8m ping

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
Ping @openssl/committers. This is a really important addition to our test capabilities.

**Review by t8m — APPROVED** [review]
Great work!

**Comment by mattcaswell** [comment]
Pushed! Thanks for the contribution. There was a commit in the history which was subsequently reverted so I removed those commits completely before pushing.
